[CAN] Enable FlexCAN for i.MX 93 EVK

CkovMk opened the pull request:
requires #69330 to be merged first.

henrikbrixandersen reviewed (changes requested):
> requires #69330 to be merged first.

How come?

CkovMk commented:
> > requires #69330 to be merged first.
> 
> How come?

This is due to the i.MX 93 EVK board design. The ADP5585 gpio expander controls:
1. A set of 1:2 digital mux that switch pin signal between 2 functions.
2. The S (or EN) pin of CAN xcvr.

JiafeiPan commented:
Please rebase this PR as https://github.com/zephyrproject-rtos/zephyr/pull/69330 has been merged.

CkovMk commented:
Rebased. note that flexan2, lpi2c2, mfd0, can_phy is enabled in board dts. will disable after ci pass.

JiafeiPan commented on soc/nxp/imx/imx9/a55/mmu_regions.c:
```diff
@@ -38,6 +38,9 @@ static const struct arm_mmu_region mmu_regions[] = {
 
+	MMU_REGION_DT_COMPAT_FOREACH_FLAT_ENTRY(nxp_flexcan,
+			 (MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS))
+
```
Would you please change the driver can_mcux_flexcan.c to use MMIO mapping (refer to https://docs.zephyrproject.org/latest/kernel/drivers/index.html#memory-mapping), so that we don't need to add MMU mapping at here.

CkovMk replied:
I do not have the bandwidth to modify flexcan driver for now... Could we left it as-is?

JiafeiPan replied:
NP, we can do it late.

CkovMk replied:
then I'll mark this resolved

JiafeiPan commented:
> Rebased. note that flexan2, lpi2c2, mfd0, can_phy is enabled in board dts. will disable after ci pass.

Is there any testing or sample application can be used to verify the driver? we can add overlay in the application and also the driver can be verified.

CkovMk commented:
not sure why ci is not building any mimx93_evk tests...

CkovMk commented:
> Is there any testing or sample application can be used to verify the driver? we can add overlay in the application and also the driver can be verified.

I wrote a test application to receive and echo can frames, and it works fine.

JiafeiPan commented:
> > Is there any testing or sample application can be used to verify the driver? we can add overlay in the application and also the driver can be verified.
> 
> I wrote a test application to receive and echo can frames, and it works fine.

Could we leverage existing application in Zephyr kernel? such as application in tests/drivers/can samples/subsys/canbus/isotp? if not upstream you test code? I mean with an application available in Kernel will make it easy to verify FlexCAN.

henrikbrixandersen commented:
Please ensure the following tests all pass with the changes from this PR:

`west twister -p imx93_evk/mimx9352/a55 -t can`

henrikbrixandersen commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.dts:
```diff
@@ -64,6 +65,14 @@
+		max-bitrate = <8000000>;
+		standby-gpios = <&gpio_exp0 8 GPIO_ACTIVE_HIGH>;
+		#phy-cells = <0>;
+		status = "disabled";
```
Why is this disabled by default?

henrikbrixandersen replied:
If they are disabled, the tests you just ran have not tested anything significant.

They need to be enabled by default if the board provides them by default. The software drivers etc. will still only be compiled in if the user/application enables `CONFIG_CAN`.

CkovMk replied:
ok. these are enabled.

henrikbrixandersen commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.dts:
```diff
@@ -145,3 +154,17 @@
+	bus-speed = <125000>;
+	bus-speed-data = <1000000>;
+	phys = <&flexcan2_phy>;
+	status = "disabled";
```
Why is this disabled by default?

henrikbrixandersen commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.dts:
```diff
@@ -145,3 +154,17 @@
+	bus-speed-data = <1000000>;
+	phys = <&flexcan2_phy>;
+	status = "disabled";
+	can-transceiver {
+		max-bitrate = <8000000>;
+	};
```
You cannot have both the `phys` property and a `can-transceiver` child node.

CkovMk replied:
removed.

henrikbrixandersen commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.yaml:
```diff
@@ -17,6 +17,7 @@ supported:
   - uart
   - i2c
   - spi
+  - can
```
This is wrong _if_ the CAN controller is disabled by default.

henrikbrixandersen commented on dts/arm64/nxp/nxp_mimx93_a55.dtsi:
```diff
@@ -322,6 +322,30 @@
+		reg = <0x443a0000 DT_SIZE_K(64)>;
+		interrupt-parent= <&gic>;
+		interrupts = <GIC_SPI 8 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>,
+		<GIC_SPI 9 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
```
Please indent this line to line up with the line above, starting with `<GIC ...`.

henrikbrixandersen commented on dts/arm64/nxp/nxp_mimx93_a55.dtsi:
```diff
@@ -322,6 +322,30 @@
+		reg = <0x425b0000 DT_SIZE_K(64)>;
+		interrupt-parent= <&gic>;
+		interrupts = <GIC_SPI 51 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>,
+		<GIC_SPI 52 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
```
Please indent this line to line up with the line above, starting with `<GIC ...`.

henrikbrixandersen commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.dts:
```diff
@@ -64,6 +65,14 @@
 		};
 	};
+
+	flexcan2_phy: can-phy-tja1057 {
```
Also, this does not follow the de-facto standard naming for CAN transceiver devicetree nodes.

```suggestion
	can_phy0: can-phy0 {
```

CkovMk commented:
> Please ensure the following tests all pass with the changes from this PR:
> 
> `west twister -p imx93_evk/mimx9352/a55 -t can`

verified pass.

```
$ west twister -p imx93_evk/mimx9352/a55 -t can
INFO    - Using Ninja..
INFO    - Zephyr version: v3.6.0-3652-g714b403b801b
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/ckovmk/Repository/zephyrproject/twister-out/testplan.json
INFO    - JOBS: 8
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete: 2581/2581  100%  skipped: 2569, failed:    0, error:    0
INFO    - 2785 test scenarios (2581 test instances) selected, 2569 configurations skipped (2568 by static filter, 1 at runtime).
INFO    - 12 of 2581 test configurations passed (100.00%), 0 failed, 0 errored, 2569 skipped with 0 warnings in 107.01 seconds
INFO    - In total 252 test cases were executed, 14114 skipped on 1 out of total 1 platforms (100.00%)
INFO    - 0 test configurations executed on platforms, 12 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/ckovmk/Repository/zephyrproject/twister-out/twister.json
INFO    - Writing xunit report /home/ckovmk/Repository/zephyrproject/twister-out/twister.xml...
INFO    - Writing xunit report /home/ckovmk/Repository/zephyrproject/twister-out/twister_report.xml...
INFO    - Run completed
```

CkovMk committed:
soc: imx93: enable flexcan driver

- Add flexcan dts node and pinctrl.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

henrikbrixandersen commented:
> verified pass.

Try again with `west twister -p imx93_evk/mimx9352/a55 -t can -v` and you'll see that no CAN controller tests are run due to the CAN devicetree nodes being disabled.

CkovMk commented:
is it the right way if I add initialize priority to board defconfig?

henrikbrixandersen commented:
> is it the right way if I add initialize priority to board defconfig?

I'd say so, yes. This is due to the GPIOs controlling the CAN transceiver coming from a GPIO expander.

CkovMk commented:
twister pass with this rev.

henrikbrixandersen commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.dts:
```diff
@@ -145,3 +154,14 @@
+&flexcan2 {
+	pinctrl-0 = <&flexcan2_default>;
+	pinctrl-names = "default";
+	sample-point = <875>;
+	sample-point-data = <875>;
```
Please remove these. This will cause the CAN timing calculations to use the default values.

henrikbrixandersen commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55_defconfig:
```diff
@@ -29,3 +29,6 @@ CONFIG_UART_CONSOLE=y
 CONFIG_CLOCK_CONTROL=y
 CONFIG_PINCTRL=y
+
+# CAN Phy must be enabled after mfd0 ready
+CONFIG_CAN_TRANSCEIVER_INIT_PRIORITY=75
```
As `CONFIG_CAN_TRANSCEIVER_INIT_PRIORITY` depends on `CONFIG_CAN`, you should move this change of default value to `boards/nxp/imx93_evk/Kconfig.defconfig` as a conditional.

CkovMk replied:
Could you please be more specific?

I tried to create `boards/nxp/imx93_evk/Kconfig.defconfig` with following content:

```c
if BOARD_IMX93_EVK_MIMX9352_A55

if CAN
# CAN Phy must be enabled after mfd0 gpio expander is ready
config CAN_TRANSCEIVER_INIT_PRIORITY
    default 75
endif

endif # BOARD_IMX93_EVK_MIMX9352_A55
```

and check the `.config` file under build directory, only to find `CONFIG_CAN_TRANSCEIVER_INIT_PRIORITY` still default to 45.

Not really that familiar with Kconfig, did I miss something?

henrikbrixandersen replied:
> and check the `.config` file under build directory, only to find `CONFIG_CAN_TRANSCEIVER_INIT_PRIORITY` still default to 45.

Did you do a completely clean build? You are only chasing the default, so rebuilding after this change will have no effect.

CkovMk replied:
after deleting build directory, still no effect.

henrikbrixandersen replied:
Since the file in question is not included in this PR, I can't really try reproducing the issue, you are seeing.

CkovMk replied:
I'll push my revision, so that you could check that.

henrikbrixandersen replied:
The changes pushed so far - are they not working?

CkovMk replied:
no, it's not.

CkovMk replied:
my bad, it is actually working...

CkovMk committed:
boards: imx93_evk: enable flexcan driver

Add pinctrl for flexcan2
Add EXP_SEL gpio hog for board mux selection during boot.
Add TJA1057 CAN PHY driver.
Update supported features in board document

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

CkovMk commented:
By running `zephyr/tests/drivers/can/api`, following errors occur:

```
*** Booting Zephyr OS build v3.6.0-3652-g12e5c8a739b2 ***
Running TESTSUITE can_classic
===================================================================
START - test_add_filter
 PASS - test_add_filter in 0.001 seconds
===================================================================
START - test_add_invalid_ext_filter
E: invalid filter with extended (29-bit) CAN ID 0x1fffffff, CAN ID mask 0x20000000
E: invalid filter with extended (29-bit) CAN ID 0x20000000, CAN ID mask 0x1fffffff
 PASS - test_add_invalid_ext_filter in 0.015 seconds
===================================================================
START - test_add_invalid_null_filter
 PASS - test_add_invalid_null_filter in 0.001 seconds
===================================================================
START - test_add_invalid_std_filter
E: invalid filter with standard (11-bit) CAN ID 0x7ff, CAN ID mask 0x800
E: invalid filter with standard (11-bit) CAN ID 0x800, CAN ID mask 0x7ff
 PASS - test_add_invalid_std_filter in 0.013 seconds
===================================================================
START - test_bitrate_limits
 PASS - test_bitrate_limits in 0.001 seconds
===================================================================
START - test_classic_get_capabilities
 PASS - test_classic_get_capabilities in 0.001 seconds
===================================================================
START - test_filters_added_while_stopped
 PASS - test_filters_added_while_stopped in 0.002 seconds
===================================================================
START - test_filters_preserved_through_bitrate_change
 PASS - test_filters_preserved_through_bitrate_change in 0.003 seconds
===================================================================
START - test_filters_preserved_through_mode_change
 PASS - test_filters_preserved_through_mode_change in 0.003 seconds
===================================================================
START - test_get_core_clock
 PASS - test_get_core_clock in 0.001 seconds
===================================================================
START - test_get_state
 PASS - test_get_state in 0.001 seconds
===================================================================
START - test_invalid_sample_point
 PASS - test_invalid_sample_point in 0.001 seconds
===================================================================
START - test_max_ext_filters
 PASS - test_max_ext_filters in 0.001 seconds
===================================================================
START - test_max_std_filters
 PASS - test_max_std_filters in 0.001 seconds
===================================================================
START - test_receive_timeout
 PASS - test_receive_timeout in 0.101 seconds
===================================================================
START - test_recover
 PASS - test_recover in 0.001 seconds
===================================================================
START - test_recover_while_stopped
 PASS - test_recover_while_stopped in 0.001 seconds
===================================================================
START - test_reject_ext_id_rtr
 PASS - test_reject_ext_id_rtr in 0.101 seconds
===================================================================
START - test_reject_std_id_rtr
 PASS - test_reject_std_id_rtr in 0.101 seconds
===================================================================
START - test_send_and_forget
 PASS - test_send_and_forget in 0.001 seconds
===================================================================
START - test_send_callback
 PASS - test_send_callback in 0.001 seconds
===================================================================
START - test_send_ext_id_out_of_range
E: invalid frame with extended (29-bit) CAN ID 0x20000000
 PASS - test_send_ext_id_out_of_range in 0.006 seconds
===================================================================
START - test_send_fd_format
E: unsupported CAN frame flags 0x04
 PASS - test_send_fd_format in 0.004 seconds
===================================================================
START - test_send_invalid_dlc
E: DLC of 9 exceeds maximum (8)
 PASS - test_send_invalid_dlc in 0.003 seconds
===================================================================
START - test_send_null_frame
 PASS - test_send_null_frame in 0.001 seconds
===================================================================
START - test_send_receive_ext_id
 PASS - test_send_receive_ext_id in 0.004 seconds
===================================================================
START - test_send_receive_ext_id_masked
 PASS - test_send_receive_ext_id_masked in 0.004 seconds
===================================================================
START - test_send_receive_ext_id_rtr
 SKIP - test_send_receive_ext_id_rtr in 0.001 seconds
===================================================================
START - test_send_receive_msgq
 PASS - test_send_receive_msgq in 0.010 seconds
===================================================================
START - test_send_receive_std_id
 PASS - test_send_receive_std_id in 0.004 seconds
===================================================================
START - test_send_receive_std_id_masked
 PASS - test_send_receive_std_id_masked in 0.004 seconds
===================================================================
START - test_send_receive_std_id_rtr
 SKIP - test_send_receive_std_id_rtr in 0.001 seconds
===================================================================
START - test_send_receive_wrong_id
 PASS - test_send_receive_wrong_id in 0.102 seconds
===================================================================
START - test_send_std_id_out_of_range
E: invalid frame with standard (11-bit) CAN ID 0x800
 PASS - test_send_std_id_out_of_range in 0.005 seconds
===================================================================
START - test_send_while_stopped
 PASS - test_send_while_stopped in 0.001 seconds
===================================================================
START - test_set_bitrate
 PASS - test_set_bitrate in 0.001 seconds
===================================================================
START - test_set_bitrate_too_high
 PASS - test_set_bitrate_too_high in 0.001 seconds
===================================================================
START - test_set_bitrate_while_started
 PASS - test_set_bitrate_while_started in 0.001 seconds
===================================================================
START - test_set_mode_while_started
 PASS - test_set_mode_while_started in 0.001 seconds
===================================================================
START - test_set_state_change_callback
 PASS - test_set_state_change_callback in 0.001 seconds
===================================================================
START - test_set_timing_max
 PASS - test_set_timing_max in 0.006 seconds
===================================================================
START - test_set_timing_min
 PASS - test_set_timing_min in 0.002 seconds
===================================================================
START - test_set_timing_while_started
 PASS - test_set_timing_while_started in 0.001 seconds
===================================================================
START - test_start_while_started
 PASS - test_start_while_started in 0.001 seconds
===================================================================
START - test_stop_while_stopped
 PASS - test_stop_while_stopped in 0.001 seconds
===================================================================
TESTSUITE can_classic succeeded
Running TESTSUITE can_stats
===================================================================
START - test_can_stats_accessors
 PASS - test_can_stats_accessors in 0.001 seconds
===================================================================
TESTSUITE can_stats succeeded
Running TESTSUITE can_transceiver
===================================================================
START - test_get_transceiver
 PASS - test_get_transceiver in 0.001 seconds
===================================================================
TESTSUITE can_transceiver succeeded
Running TESTSUITE can_utilities
===================================================================
START - test_can_bytes_to_dlc
 PASS - test_can_bytes_to_dlc in 0.001 seconds
===================================================================
START - test_can_dlc_to_bytes
 PASS - test_can_dlc_to_bytes in 0.001 seconds
===================================================================
START - test_can_frame_matches_filter
 PASS - test_can_frame_matches_filter in 0.001 seconds
===================================================================
TESTSUITE can_utilities succeeded
Running TESTSUITE canfd
===================================================================
START - test_canfd_get_capabilities
 PASS - test_canfd_get_capabilities in 0.001 seconds
===================================================================
START - test_filters_preserved_through_classic_to_fd_mode_change

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/can/api/src/canfd.c:101: send_test_frame: (err not equal to 0)
failed to send frame (err -114)
 FAIL - test_filters_preserved_through_classic_to_fd_mode_change in 0.015 seconds
===================================================================
START - test_filters_preserved_through_fd_to_classic_mode_change

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/can/api/src/canfd.c:101: send_test_frame: (err not equal to 0)
failed to send frame (err -114)
 FAIL - test_filters_preserved_through_fd_to_classic_mode_change in 0.015 seconds
===================================================================
START - test_invalid_sample_point
 PASS - test_invalid_sample_point in 0.001 seconds
===================================================================
START - test_send_receive_classic

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/can/api/src/canfd.c:161: add_rx_filter: (filter_id equal to -ENOSPC)
no filters available
 FAIL - test_send_receive_classic in 0.014 seconds
===================================================================
START - test_send_receive_fd

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/can/api/src/canfd.c:137: add_rx_msgq: (filter_id equal to -ENOSPC)
no filters available
 FAIL - test_send_receive_fd in 0.013 seconds
===================================================================
START - test_send_receive_mixed

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/can/api/src/canfd.c:137: add_rx_msgq: (filter_id equal to -ENOSPC)
no filters available
 FAIL - test_send_receive_mixed in 0.013 seconds
===================================================================
START - test_set_bitrate_data_while_started
 PASS - test_set_bitrate_data_while_started in 0.001 seconds
===================================================================
START - test_set_bitrate_too_high
 PASS - test_set_bitrate_too_high in 0.001 seconds
===================================================================
START - test_set_timing_data_max
 PASS - test_set_timing_data_max in 0.001 seconds
===================================================================
START - test_set_timing_data_min
 PASS - test_set_timing_data_min in 0.001 seconds
===================================================================
START - test_set_timing_data_while_started
 PASS - test_set_timing_data_while_started in 0.001 seconds
===================================================================
TESTSUITE canfd failed.

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [can_classic]: pass = 43, fail = 0, skip = 2, total = 45 duration = 0.518 seconds
 - PASS - [can_classic.test_add_filter] duration = 0.001 seconds
 - PASS - [can_classic.test_add_invalid_ext_filter] duration = 0.015 seconds
 - PASS - [can_classic.test_add_invalid_null_filter] duration = 0.001 seconds
 - PASS - [can_classic.test_add_invalid_std_filter] duration = 0.013 seconds
 - PASS - [can_classic.test_bitrate_limits] duration = 0.001 seconds
 - PASS - [can_classic.test_classic_get_capabilities] duration = 0.001 seconds
 - PASS - [can_classic.test_filters_added_while_stopped] duration = 0.002 seconds
 - PASS - [can_classic.test_filters_preserved_through_bitrate_change] duration = 0.003 seconds
 - PASS - [can_classic.test_filters_preserved_through_mode_change] duration = 0.003 seconds
 - PASS - [can_classic.test_get_core_clock] duration = 0.001 seconds
 - PASS - [can_classic.test_get_state] duration = 0.001 seconds
 - PASS - [can_classic.test_invalid_sample_point] duration = 0.001 seconds
 - PASS - [can_classic.test_max_ext_filters] duration = 0.001 seconds
 - PASS - [can_classic.test_max_std_filters] duration = 0.001 seconds
 - PASS - [can_classic.test_receive_timeout] duration = 0.101 seconds
 - PASS - [can_classic.test_recover] duration = 0.001 seconds
 - PASS - [can_classic.test_recover_while_stopped] duration = 0.001 seconds
 - PASS - [can_classic.test_reject_ext_id_rtr] duration = 0.101 seconds
 - PASS - [can_classic.test_reject_std_id_rtr] duration = 0.101 seconds
 - PASS - [can_classic.test_send_and_forget] duration = 0.001 seconds
 - PASS - [can_classic.test_send_callback] duration = 0.001 seconds
 - PASS - [can_classic.test_send_ext_id_out_of_range] duration = 0.006 seconds
 - PASS - [can_classic.test_send_fd_format] duration = 0.004 seconds
 - PASS - [can_classic.test_send_invalid_dlc] duration = 0.003 seconds
 - PASS - [can_classic.test_send_null_frame] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_ext_id] duration = 0.004 seconds
 - PASS - [can_classic.test_send_receive_ext_id_masked] duration = 0.004 seconds
 - SKIP - [can_classic.test_send_receive_ext_id_rtr] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_msgq] duration = 0.010 seconds
 - PASS - [can_classic.test_send_receive_std_id] duration = 0.004 seconds
 - PASS - [can_classic.test_send_receive_std_id_masked] duration = 0.004 seconds
 - SKIP - [can_classic.test_send_receive_std_id_rtr] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_wrong_id] duration = 0.102 seconds
 - PASS - [can_classic.test_send_std_id_out_of_range] duration = 0.005 seconds
 - PASS - [can_classic.test_send_while_stopped] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate_too_high] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_set_mode_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_set_state_change_callback] duration = 0.001 seconds
 - PASS - [can_classic.test_set_timing_max] duration = 0.006 seconds
 - PASS - [can_classic.test_set_timing_min] duration = 0.002 seconds
 - PASS - [can_classic.test_set_timing_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_start_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_stop_while_stopped] duration = 0.001 seconds

SUITE PASS - 100.00% [can_stats]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.001 seconds
 - PASS - [can_stats.test_can_stats_accessors] duration = 0.001 seconds

SUITE PASS - 100.00% [can_transceiver]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.001 seconds
 - PASS - [can_transceiver.test_get_transceiver] duration = 0.001 seconds

SUITE PASS - 100.00% [can_utilities]: pass = 3, fail = 0, skip = 0, total = 3 duration = 0.003 seconds
 - PASS - [can_utilities.test_can_bytes_to_dlc] duration = 0.001 seconds
 - PASS - [can_utilities.test_can_dlc_to_bytes] duration = 0.001 seconds
 - PASS - [can_utilities.test_can_frame_matches_filter] duration = 0.001 seconds

SUITE FAIL -  58.33% [canfd]: pass = 7, fail = 5, skip = 0, total = 12 duration = 0.077 seconds
 - PASS - [canfd.test_canfd_get_capabilities] duration = 0.001 seconds
 - FAIL - [canfd.test_filters_preserved_through_classic_to_fd_mode_change] duration = 0.015 seconds
 - FAIL - [canfd.test_filters_preserved_through_fd_to_classic_mode_change] duration = 0.015 seconds
 - PASS - [canfd.test_invalid_sample_point] duration = 0.001 seconds
 - FAIL - [canfd.test_send_receive_classic] duration = 0.014 seconds
 - FAIL - [canfd.test_send_receive_fd] duration = 0.013 seconds
 - FAIL - [canfd.test_send_receive_mixed] duration = 0.013 seconds
 - PASS - [canfd.test_set_bitrate_data_while_started] duration = 0.001 seconds
 - PASS - [canfd.test_set_bitrate_too_high] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_max] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_min] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_while_started] duration = 0.001 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION FAILED
```

It seems to be working fine in classic can, failing can-fd tests. I don't think it is related to soc & board level enablement in this pr.
I'll open an issue for that.

Ping @dleach02  @henrikbrixandersen  @JiafeiPan

CkovMk commented:
issue: #73757

henrikbrixandersen commented:
> By running `zephyr/tests/drivers/can/api`, following errors occur:

How does this correlate with your comment https://github.com/zephyrproject-rtos/zephyr/pull/70201#issuecomment-2102171581 here saying all tests pass?

CkovMk commented:
> > By running `zephyr/tests/drivers/can/api`, following errors occur:
> 
> How does this correlate with your comment [#70201 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/70201#issuecomment-2102171581) here saying all tests pass?

Sorry, I was meaning all test build pass. But that's not the case here either. Let me check what goes wrong. It seems what caused the build failure is some nodes are disabled / KConfkg not enabled.

JiafeiPan commented:
> > > By running `zephyr/tests/drivers/can/api`, following errors occur:
> > 
> > 
> > How does this correlate with your comment [#70201 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/70201#issuecomment-2102171581) here saying all tests pass?
> 
> Sorry, I was meaning all test build pass. But that's not the case here either. Let me check what goes wrong. It seems what caused the build failure is some nodes are disabled / KConfkg not enabled.

As we discussed, you should disable related CAN nodes in default dts, and enable them in overlay of testcase.

henrikbrixandersen commented:
> As we discussed, you should disable related CAN nodes in default dts, and enable them in overlay of testcase.

What leads you to believe this is the right solution? This is the opposite of what all other boards do.

JiafeiPan commented:
> > As we discussed, you should disable related CAN nodes in default dts, and enable them in overlay of testcase.
> 
> What leads you to believe this is the right solution? This is the opposite of what all other boards do.

Would you please give more detailed explanation why enable all the device nodes in board dts is the best choice?  For MPU platform, we need to run multiple OS on the single platform,  so that we can't enable the same peripherals on multiple OS, that is to say, in order to run a hello world zephyr simultaneously with running Linux, we have to disable so much unused dts nodes in Zephyr board dts as all the peripherals have already be enabled in Linux, right?

henrikbrixandersen commented:
> Would you please give more detailed explanation why enable all the device nodes in board dts is the best choice? For MPU platform, we need to run multiple OS on the single platform, so that we can't enable the same peripherals on multiple OS, that is to say, in order to run a hello world zephyr simultaneously with running Linux, we have to disable so much unused dts nodes in Zephyr board dts as all the peripherals have already be enabled in Linux, right?

Enabling a devicetree node does not enable the peripheral driver unless the corresponding Kconfig is enabled as well. So in your example, the hello world Zephyr application would not enable `CONFIG_CAN` and therefore enabling the FlexCAN devicetree node has no impact.

JiafeiPan commented:
> > Would you please give more detailed explanation why enable all the device nodes in board dts is the best choice? For MPU platform, we need to run multiple OS on the single platform, so that we can't enable the same peripherals on multiple OS, that is to say, in order to run a hello world zephyr simultaneously with running Linux, we have to disable so much unused dts nodes in Zephyr board dts as all the peripherals have already be enabled in Linux, right?
> 
> Enabling a devicetree node does not enable the peripheral driver unless the corresponding Kconfig is enabled as well. So in your example, the hello world Zephyr application would not enable `CONFIG_CAN` and therefore enabling the FlexCAN devicetree node has no impact.

Oh.., my bad, clear now, thanks for your explanation.
